### PR TITLE
[CodeCompletion] Add 'IsSystem' flag to code completion result item

### PR DESCRIPTION
**Pull request by rintaro**
`key.is_system: 1` is added if the associated declaration is from a system module.

rdar://problem/62617558

(This will conflict with #31637, I will rebase it after merging this)

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please test

File: include/swift/IDE/CodeCompletion.h [hunk 1]
```diff
@@ -594,14 +594,15 @@ class CodeCompletionResult {
...
 
   /// The number of bytes to the left of the code completion point that
   /// should be erased first if this completion string is inserted in the
   /// editor buffer.
-  unsigned NumBytesToErase : 7;
+  unsigned NumBytesToErase : 6;
 
 public:
-  static const unsigned MaxNumBytesToErase = 127;
+  static const unsigned MaxNumBytesToErase = 63;
```
Comment (rintaro) [inline include/swift/IDE/CodeCompletion.h:605]: I think 63 is still more than enough.

**Comment by rintaro** [comment]
@swift-ci Please test

File: include/swift/IDE/CodeCompletion.h (hunk 1 above)
Comment (akyrtzi) [reply include/swift/IDE/CodeCompletion.h:605]: @benlangmuir WDYT?

Comment (rintaro) [reply include/swift/IDE/CodeCompletion.h]: As per offline discussion, we decided to keep it 127.

File: tools/SourceKit/tools/sourcekitd/lib/API/CodeCompletionResultsArray.cpp [hunk 2]
```diff
@@ -56,10 +56,15 @@ void CodeCompletionResultsArrayBuilder::add(
...
+    bool IsSystem,
     unsigned NumBytesToErase) {
 
-  assert(NumBytesToErase <= (uint8_t(-1) >> 1));
-  uint8_t BytesAndNotRecommended = (NumBytesToErase << 1) | NotRecommended;
+  assert(NumBytesToErase <= (unsigned(-1) << 2));
+  unsigned BytesAndFlags = 0;
+  BytesAndFlags |= NumBytesToErase << 2;
+  BytesAndFlags |= NotRecommended << 1;
+  BytesAndFlags |= IsSystem << 0;
```
Comment (rintaro) [inline tools/SourceKit/tools/sourcekitd/lib/API/CodeCompletionResultsArray.cpp:66]: I just noticed each value is just casted to pointer size:
https://github.com/apple/swift/blob/eaa1051dfa269a3b6fe7418796140f9f313480c0/tools/SourceKit/tools/sourcekitd/lib/API/CompactArray.cpp#L22-L39
So I decided to use `unsigned` as a bitfield.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/IDE/CodeCompletion.cpp [hunk 3]
```diff
@@ -4714,9 +4720,11 @@ class CompletionOverrideLookup : public swift::VisibleDeclConsumer {
...
         auto dist = Ctx.SourceMgr.getByteDistance(
             introducerLoc, Ctx.SourceMgr.getCodeCompletionLoc());
-        Builder.setNumBytesToErase(dist);
-        Builder.addOverrideKeyword();
-        Builder.addDeclIntroducer(DeclStr.str().substr(0, NameOffset));
+        if (dist <= CodeCompletionResult::MaxNumBytesToErase) {
+          Builder.setNumBytesToErase(dist);
+          Builder.addOverrideKeyword();
+          Builder.addDeclIntroducer(DeclStr.str().substr(0, NameOffset));
+        }
```
Comment (rintaro) [inline lib/IDE/CodeCompletion.cpp:4727]: @akyrtzi There was no `MaxNumBytesToErase` for override completion.

If the `dist` exceed `MaxNumBytesToErase`, completion just gives up adding `override` keyword before the introducer (e.g. `func`, `var`).

**Review by akyrtzi — APPROVED** [review]

File: test/SourceKit/CodeComplete/complete_from_system.swift [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+func test(col: MyCollection) {
+  col.
+}
+
+// RUN: %sourcekitd-test -req=complete -pos=9:7 %s -- %s -module-name TestMod | %FileCheck %s
+
+// CHECK: {
+// CHECK: key.kind: source.lang.swift.decl.function.method.instance,
+// CHECK: key.name: "makeIterator()",
+// CHECK-NEXT: key.sourcetext: "makeIterator()",
```
Comment (benlangmuir) [inline test/SourceKit/CodeComplete/complete_from_system.swift:17]: Do we really care about testing all these fields here? Can we just check is_system specifically?

Comment (rintaro) [reply test/SourceKit/CodeComplete/complete_from_system.swift:17]: Is there a way to check that the line is in the same item (i.e. before `},` line )?

Comment (benlangmuir) [reply test/SourceKit/CodeComplete/complete_from_system.swift:17]: It's not a great solution, but

```
CHECK: key.name: "blah blah blah"
CHECK-NOT: key.name
CHECK: key.is_system: 1
```

Comment (rintaro) [reply test/SourceKit/CodeComplete/complete_from_system.swift:17]: Cool, that actually works for `is_system: 1` checks. I still haven't found a way to check the item does *not* have `is_system: 1` line.

Comment (rintaro) [reply test/SourceKit/CodeComplete/complete_from_system.swift:17]:
```
// CHECK: key.name: "makeIterator()"
// CHECK-NOT: }
// CHECK: key.is_system: 1
// CHECK: }

// CHECK: key.name: "startIndex"
// CHECK-NOT: }
// CHECK-NOT: key.is_system: 1
// CHECK: },
```
This seems to work.

Comment (benlangmuir) [reply test/SourceKit/CodeComplete/complete_from_system.swift:17]: Only danger is that we could add a dictionary in the middle of the response some day, but not sure there is a better solution right now.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by benlangmuir — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

**Pull request merged**